codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java

```diff
@@ -0,0 +1,824 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+import io.netty.handler.codec.DefaultHeaders;
+import io.netty.handler.codec.Headers;
+import io.netty.util.AsciiString;
+import io.netty.util.internal.PlatformDependent;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static io.netty.handler.codec.http2.DefaultHttp2Headers.HTTP2_NAME_VALIDATOR;
+import static io.netty.util.internal.EmptyArrays.EMPTY_ASCII_STRINGS;
+
+public final class ReadOnlyHttp2Headers implements Http2Headers {
+    private final AsciiString[] pseudoHeaders;
+    private final AsciiString[] otherHeaders;
+
+    /**
+     * Used to create read only object designed to represent trailers.
+     * <p>
+     * If this is used for a purpose other than trailers you may violate the header serialization ordering defined by
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.1">RFC 7540, 8.1.2.1</a>.
+     * @param otherHeaders A an array of key:value pairs. Must not contain any
+     *                     <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.1">pseudo headers</a>
+     *                     or {@code null} names/values.
+     * @param validateHeaders {@code true} will run validation on each header name/value pair to ensure protocol
+     *                        compliance.
+     * @return A read only representation of the headers.
+     */
+    public static ReadOnlyHttp2Headers trailers(AsciiString[] otherHeaders, boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(EMPTY_ASCII_STRINGS, otherHeaders, validateHeaders);
```

Comment:
And the null gaps?  gRPC's Headers used nulls to indicate as a shortcut that there were no more headers in the array.  Jakob added this, but that's because we could be really flexible with it.  Here, I think you still need to call it out.